src/main/java/com/couchbase/lite/replicator/PullerInternal.java

```diff
@@ -456,10 +459,27 @@ private void queueDownloadedRevision(RevisionInternal rev) {
             }
         }
 
-        if (rev != null && rev.getBody() != null)
-            rev.getBody().compact();
+        final Body body = rev.getBody();
+        if (rev != null && body != null)
+            body.compact();
+
+        synchronized (downloadsToInsert) {
+            globalMemorySize+=body.size();
+        }
 
         downloadsToInsert.queueObject(rev);
+        if (tooManyQueuedObjects()) {
+            Log.d(TAG, "Flushing queue global memory size at " + globalMemorySize);
+            synchronized (downloadsToInsert) {
+                downloadsToInsert.flushAll(true);
+                Log.d(TAG, "Resetting global memory size.");
+                globalMemorySize=0;
+            }
+        }
+    }
+
+    private boolean tooManyQueuedObjects() {
+        return globalMemorySize > MEGABYTE * 0.5;
```

Comment:
Agreed, it's an arbitrary number. I tried experimenting with 1MB at first but was still getting OOMemory though it happens less frequently with more aggressive flushing.